File: firmware/ghidra_scripts/dump_re.py
# dump_re.py — Ghidra Jython post-analysis dump for the DTX-MULTI 12 RE.
# Run headless via: analyzeHeadless <proj> M12 -process code.bin -postScript dump_re.py
# Writes a text report next to the firmware.
from ghidra.app.decompiler import DecompInterface
from ghidra.util.task import ConsoleTaskMonitor

OUT = "/Users/velvetmotion/Desktop/DTX-M12/firmware/ghidra_dump.txt"
fm  = currentProgram.getFunctionManager()
af  = currentProgram.getAddressFactory()
rm  = currentProgram.getReferenceManager()
def A(x): return af.getAddress(hex(x).rstrip("L"))

lines = []
def out(s): lines.append(s)

out("function count: %d" % fm.getFunctionCount())

# 1) Who references the FX descriptor table region and the label pool?
def refs_to(a):
    res = {}
    for r in rm.getReferencesTo(A(a)):
        f = fm.getFunctionContaining(r.getFromAddress())
        if f: res[f.getEntryPoint().getOffset()] = f.getName()
    return res

for name, a in (("FX_descriptor_table ~0x0C096C00", 0x0C096C00),
                ("label_pool 0x0C08DFE8", 0x0C08DFE8),
                ("label_pool_mid 0x0C08E404", 0x0C08E404)):
    out("\n=== refs to %s ===" % name)
    for off, nm in sorted(refs_to(a).items()):
        out("  %s @ 0x%08X" % (nm, off))

# 2) Decompile target functions (FX handlers + any callers we discover).
dec = DecompInterface(); dec.openProgram(currentProgram)
mon = ConsoleTaskMonitor()
def decompile(a, tag=""):
    f = fm.getFunctionContaining(A(a))
    if not f:
        out("\n--- no function at 0x%08X %s ---" % (a, tag)); return
    out("\n===== %s  %s @ %s =====" % (tag, f.getName(), f.getEntryPoint()))
    r = dec.decompileFunction(f, 90, mon)
    if r and r.getDecompiledFunction():
        out(r.getDecompiledFunction().getC())
    else:
        out("  <decompile failed>")

for a in (0x0C0217EA, 0x0C021832, 0x0C0218C0, 0x0C023DB0):
    decompile(a, "FXhandler")

open(OUT, "w").write("\n".join(lines))
print("dump_re: wrote %s (%d lines)" % (OUT, len(lines)))

File: firmware/ghidra_scripts/DumpRE.java
// DumpRE.java — headless post-analysis dump for the DTX-MULTI 12 RE.
// Run: analyzeHeadless <proj> M12 -process code.bin -noanalysis \
//        -scriptPath <dir> -postScript DumpRE.java
import ghidra.app.script.GhidraScript;
import ghidra.app.decompiler.*;
import ghidra.program.model.listing.*;
import ghidra.program.model.address.Address;
import ghidra.program.model.symbol.*;
import ghidra.util.task.ConsoleTaskMonitor;
import java.io.PrintWriter;

public class DumpRE extends GhidraScript {
    public void run() throws Exception {
        StringBuilder sb = new StringBuilder();
        FunctionManager fm = currentProgram.getFunctionManager();
        ReferenceManager rm = currentProgram.getReferenceManager();
        sb.append("function count: " + fm.getFunctionCount() + "\n");

        long[] refTargets = {0x0C096C00L, 0x0C08DFE8L, 0x0C08E404L, 0x0C096C48L};
        for (long a : refTargets) {
            sb.append("\n=== refs to 0x" + Long.toHexString(a) + " ===\n");
            for (Reference r : rm.getReferencesTo(toAddr(a))) {
                Function f = fm.getFunctionContaining(r.getFromAddress());
                sb.append("  from " + r.getFromAddress() + " " + r.getReferenceType()
                          + (f != null ? (" in " + f.getName() + "@" + f.getEntryPoint()) : "") + "\n");
            }
        }

        DecompInterface dec = new DecompInterface();
        dec.openProgram(currentProgram);
        ConsoleTaskMonitor mon = new ConsoleTaskMonitor();
        // mix of the small "stub" pointers and the larger handlers seen in the FX table
        long[] funcs = {0x0C021712L, 0x0C02171AL, 0x0C0217EAL, 0x0C021832L, 0x0C0218C0L, 0x0C023DB0L};
        for (long a : funcs) {
            Function f = fm.getFunctionContaining(toAddr(a));
            if (f == null) { sb.append("\n--- no function at 0x" + Long.toHexString(a) + " ---\n"); continue; }
            sb.append("\n===== " + f.getName() + " @ " + f.getEntryPoint()
                      + "  (size " + f.getBody().getNumAddresses() + ") =====\n");
            DecompileResults r = dec.decompileFunction(f, 120, mon);
            if (r != null && r.getDecompiledFunction() != null)
                sb.append(r.getDecompiledFunction().getC());
            else
                sb.append("  <decompile failed>\n");
        }

        PrintWriter pw = new PrintWriter("/Users/velvetmotion/Desktop/DTX-M12/firmware/ghidra_dump.txt");
        pw.print(sb.toString());
        pw.close();
        println("DumpRE wrote ghidra_dump.txt (" + sb.length() + " chars)");
    }
}
